connection_pool/round_robin.go

```diff
@@ -60,58 +61,40 @@ func (r *RoundRobinStrategy) IsEmpty() bool {
 	return r.size == 0
 }
 
-func (r *RoundRobinStrategy) CloseConns() []error {
-	r.mutex.Lock()
-	defer r.mutex.Unlock()
-
-	errs := make([]error, len(r.conns))
-
-	for i, conn := range r.conns {
-		errs[i] = conn.Close()
-	}
-
-	return errs
-}
-
 func (r *RoundRobinStrategy) GetNextConnection() *tarantool.Connection {
 	r.mutex.RLock()
 	defer r.mutex.RUnlock()
 
-	// We want to iterate through the elements in a circular order
-	// so the first element in cycle is connections[next]
-	// and the last one is connections[next + length].
-	next := r.nextIndex()
-	cycleLen := len(r.conns) + next
-	for i := next; i < cycleLen; i++ {
-		idx := i % len(r.conns)
-		if r.conns[idx].ConnectedNow() {
-			if i != next {
-				atomic.StoreUint64(&r.current, uint64(idx))
-			}
-			return r.conns[idx]
-		}
+	if r.size == 0 {
+		return nil
 	}
-
-	return nil
+	return r.conns[r.nextIndex()]
 }
 
 func NewEmptyRoundRobin(size int) *RoundRobinStrategy {
 	return &RoundRobinStrategy{
 		conns:       make([]*tarantool.Connection, 0, size),
-		indexByAddr: make(map[string]int),
+		indexByAddr: make(map[string]uint),
 		size:        0,
+		current:     0,
 	}
 }
 
 func (r *RoundRobinStrategy) AddConn(addr string, conn *tarantool.Connection) {
 	r.mutex.Lock()
 	defer r.mutex.Unlock()
 
-	r.conns = append(r.conns, conn)
-	r.indexByAddr[addr] = r.size
-	r.size += 1
+	if idx, ok := r.indexByAddr[addr]; ok {
+		r.conns[idx] = conn
```

Comment:
Why do we replace connection in array? Don't we need to close old connection in this case?

Reply:
> Why do we replace connection in array?

This is debatable, but for my taste it is no worse than other approaches. We act here like a regular map.

> Don't we need to close old connection in this case?

It seems to me that this is not the responsibility of the container/strategy for choosing a next connection. This should be done by a code on a upper level of the call stack.

Reply:
> It seems to me that this is not the responsibility of the container/strategy for choosing a next connection. This should be done by a code on a upper level of the call stack.

I agree this is not the responsibility of the container/strategy, but we open connections by `adds` in  `fillPools` -> `setConnectionToPool` when we `Connect`

https://github.com/tarantool/go-tarantool/blob/a8e03440f3c74ad6e593cc327e36ce9aba8f2a90/connection_pool/connection_pool.go#L730

So, I guess it is possible we create connection in `fillPools`, store it, then got duplicate and just replace connection in container without closing the previous one on `fillPools`/`setConnectionToPool` level. Don't we need to close previous connection on `fillPools`/`setConnectionToPool` level?

Maybe I am wrong, I'm trying to figure out where I'm missing something, sorry for bothering you.

Reply:
> So, I guess it is possible we create connection in `fillPools`, store it, then got duplicate and just replace connection in container without closing the previous one on `fillPools`/`setConnectionToPool` level. Don't we need to close previous connection on `fillPools`/`setConnectionToPool` level?

Do you mean that we may have duplicates in options? Yes, it is a problem. It looks like we don't have to hold duplicates.

I've added a commit:
`bugfix: prevent duplicate connects in pool`

Reply:
Thank you!